twister: Twister cannot properly handle runners errors (flashing)

**Describe the bug**
Errors occurring during flashing devices are not handled properly. There are several scenarios related to this issue. 
I am able to reliably reproduce it on nrf5340 board, which raises errors during programming when app protected is not disabled, however, I see the same issues occasionally on other boards when an error occurs during programming. 
Scenarios:
1.) configuration: when '--ninja' is used (both with and without `--hardware-map`): the test fails as soon as there is an error (which is good). However, the fail is reported as `FAILED Cmake build failure (device)` and treated as build error. Therefore, using `--retry-failed` does not retries such failures (which is bad). 
- expected behavior: Such fail is not treated as a build error and the failed test is retried when `--retry-failed` is used

2.) configuration: without `--ninja` and without `--hardware-map`: the test ends only after timeout, even though "make" reports earlier there was a fail (`Makefile:748: recipe for target 'flash' failed`). This adds ~60s of wasted time to the test suite which can stack to some meaningful time in big test suites. In this scenario, `--retry-failed` works as it should 
- expected behavior:  test should be terminated as soon as there is an error in make for flash.

3.) configuration: without `--ninja` and with `--hardware-map`: Similar to 2.). This time, the output about `make flash` failing is not printed out. The rest is the same. Tests can be retried but are terminated only by timeouts. 
- expected behavior:  test should be terminated as soon as there is an error in make for flash.

The fact that hardware-related errors are seen as build errors results in printing out/saving in the report a lot of useless information which hinders the real issues. In such a case the full build log is dumped, even though the building was fine. 

**Impact**
Can significantly increase the execution time for big test suites. Failed tests might not get retried. Hinders real errors with full build logs. 


**Environment (please complete the following information):**
 - OS Ubuntu: 18.04
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: zephyr-v2.5.0-276-g8e75fff5808f


**Additional context**
@nashif I also wonder if this could be the cause for "log mixing" problem #25719. I saw in my setups, that when flashing failed, the already programmed app can be executed giving not matching output.

--- a/scripts/pylib/twister/twisterlib.py
+++ b/scripts/pylib/twister/twisterlib.py
@@ -814,6 +814,8 @@ class DeviceHandler(Handler):
                         self.instance.reason = "Device issue (Flash?)"
                         with open(d_log, "w") as dlog_fp:
                             dlog_fp.write(stderr.decode())
+                        os.write(write_pipe, b'x')  # halt the thread
+                        out_state = "flash_error"
                 except subprocess.TimeoutExpired:
                     proc.kill()
                     (stdout, stderr) = proc.communicate()
@@ -846,12 +848,15 @@ class DeviceHandler(Handler):
 
         handler_time = time.time() - start_time
 
-        if out_state == "timeout":
+        if out_state in ["timeout", "flash_error"]:
             for c in self.instance.testcase.cases:
                 if c not in harness.tests:
                     harness.tests[c] = "BLOCK"
 
-            self.instance.reason = "Timeout"
+            if out_state == "timeout":
+                self.instance.reason = "Timeout"
+            elif out_state == "flash_error":
+                self.instance.reason = "Flash error"
 
         self.instance.results = harness.tests
 
@@ -864,7 +869,7 @@ class DeviceHandler(Handler):
 
         if harness.state:
             self.set_state(harness.state, handler_time)
-            if  harness.state == "failed":
+            if harness.state == "failed":
                 self.instance.reason = "Failed"
         else:
             self.set_state(out_state, handler_time)
@@ -2411,7 +2416,7 @@ class ProjectBuilder(FilterBuilder):
         results.done += 1
         instance = self.instance
 
-        if instance.status in ["error", "failed", "timeout"]:
+        if instance.status in ["error", "failed", "timeout", "flash_error"]:
             if instance.status == "error":
                 results.error += 1
             results.failed += 1
@@ -3371,7 +3376,7 @@ class TestSuite(DisablePyTestCollectionMixin):
                         else:
                             fails += 1
                 else:
-                    if instance.status in ["error", "failed", "timeout"]:
+                    if instance.status in ["error", "failed", "timeout", "flash_error"]:
                         if instance.reason in ['build_error', 'handler_crash']:
                             errors += 1
                         else:
@@ -3457,7 +3462,7 @@ class TestSuite(DisablePyTestCollectionMixin):
                                     eleTestcase,
                                     'error',
                                     type="failure",
-                                    message="failed")
+                                    message=instance.reason)
                             log_root = os.path.join(self.outdir, instance.platform.name, instance.testcase.name)
                             log_file = os.path.join(log_root, "handler.log")
                             el.text = self.process_log(log_file)
@@ -3488,7 +3493,7 @@ class TestSuite(DisablePyTestCollectionMixin):
                         name="%s" % (instance.testcase.name),
                         time="%f" % handler_time)
 
-                    if instance.status in ["error", "failed", "timeout"]:
+                    if instance.status in ["error", "failed", "timeout", "flash_error"]:
                         failure = ET.SubElement(
                             eleTestcase,
                             'failure',

